### DISCUSSION TITLE
Process Behaviour Charts in FlowViz

### DISCUSSION THREAD

**Discussion by huserben** [nbrown02/FlowViz#112] [Ideas]
Hi @nbrown02 

I'm reading Dan Vacanti's new book [Actionable Agile Metrics Volume II](https://leanpub.com/actionableagilemetricsii). The main topic is Process Behaviour Charts (PBCs), or more specifically, XmR charts.

Would this/could be something to add to the FlowViz template?

As I was curious and wanted to play around with it, I used the template as a base and built a "quick and dirty" version for one of my teams which shows XmR charts for the four metrics of flow:

![XmR Overview](https://github.com/nbrown02/FlowViz/assets/5486874/b5aaf0fc-2b1d-4a76-82c2-52b72faf463c)

While it would need a bit of polishing (and someone that checks if I did the calculations correctly), and certainly some guidance from your side on how to properly add this to the template, I think it could be a good addition.

Let me know if this would be of interest, and we can sync up on how to best proceed.

Cheers
Benji

**Comment by nbrown02** [comment]
Funny you say this - a few months back I started with a version of this for ADO which I tweaked after reading Dans book:

https://github.com/nbrown02/Process-Behaviour-Chart

It is just X charts atm but mR could easily be added. I do have plans in the new year to add it to FlowViz :)

**Reply by nbrown02** [reply]
One difference with this template is it asks for an additional parameter of Team - loading this for an ADO project (with multiple teams) is where it could get messy...but I have some ideas on how to address this...

**Reply by huserben** [reply]
Great, I guess I'm late to the party discovering PBCs 😄 

The template looks great (I'm not so good a the visualizing part, not my cup of tea and still learning Power BI). I'm curious, what is the reason you additionally need the team name?

Another thought I had: Does it make sense to include it into FlowViz or might it get a bit too "cluttered"? As I could live happily with having 2 distinct reports.

If there is anything I could do to help, please let me know. Happy to support you!

**Reply by huserben** [reply]
Some first thoughts after playing around. Please keep in mind I'm still learning about those charts, so what I'm saying might not make any sense (sorry for that in advance).

- Would it make sense to be able to show/hide the 2/1 sigma lines? In my Power BI, I showed them, and I know there are rules based on those lines (that could be used instead of the 25% rule you mentioned in the readme). On the other hand, it can be a bit messy if we get 7 lines per chart...
- Am I assuming correctly that the averages and U/L NPLs are recalculated whenever the data is refreshed? Should this not be "static" once we choose a value for those and only "sporadically" be updated? For example, when we significantly changed the process?
- Lastly, I was wondering why the Throughput is sampled "per Week", while the others (WIP/WIA) are "per day"? To match with the current display in FlowViz or is there another reason?

**Reply by nbrown02** [reply]
Nothing you're saying so far is wrong :) 

- Yeah that's why I created a separate template for now as I did think it made the original FlowViz a bit cluttered :) 

- I use a team filter as it just reduces the amount of data needed to load thus a better UX plus, at the time, I wasn't sure if loading all teams within an ADO project was the best idea.

- I didn't include the sigma lines as I think it looks messy with them plus I think the majority of PBCs you see they just have the UNPL/LNPL lines :)

- Yes you are correct, they are calculated depending on the range you select. It might be nice to have two filters (one for choosing the baseline, the other for your actual data) - is that what you mean?

- Yeah you are correct on this too :) it was just because the current TP in FlowViz is weekly. As this is the input into the Monte Carlo sims (in FlowViz) it allows you to see the predictability of your input data for your simulations...

**Reply by huserben** [reply]
Thanks for answering all my questions. I'm looking forward to using the template once I'm back at work next year.

Regarding the baseline, I quote from Dan's book (literally just read this yesterday, so I had it fresh on my mind):
> Once you have established a baseline period that characterizes your process, those limits should generally remain static even as new data is added. But they don’t necessarily remain static indefinitely. New limits should be calculated whenever a process change is made and/or whenever you feel like your baseline data has become stale.

This is what I referred to. So I was wondering if there should be a "static" one, but I do admit it's harder to use then. When I was thinking about the whole thing in the morning, I had in mind to have some kind of "button" that would say "Recalculate" and then set a new baseline and U/L NPLs based on the selected data. But I don't know if that's even possible with Power BI...

"As this is the input into the Monte Carlo sims (in FlowViz) it allows you to see the predictability of your input data for your simulations..."
--> Oh right, I was not even thinking about this before. But it means if our TP is within the UNPL and LNPL, the MC Forecast should be more "trustworthy"?

**Reply by nbrown02** [reply]
- Yeah it becomes quite tricky to build something that way. My idea would be to have a slider to "select baseline data" and then another to "show last x weeks" - this way the user gets the latest data but can control the baseline themselves

- Precisely that yes, and for single item forecasts if any metrics like WIP/Cycle Time/Age are outside the UNPL/LNPL then these forecasts are also to be taken with a pinch of salt. Prateek makes the argument that if total age is within limits then the other two measures "should" also be within it (I have a blog on this currently in internal review (not sure if the link will work/might need you to sign in to view - https://nbrown02.medium.com/469c654fdbcb)

**Reply by huserben** [reply]
- Ok that makes sense, I think that would be a nice compromise
- Great post, thanks for sharing.

Question: Should the discussion be closed here, or shall it be left open? Maybe in the readme of FlowViz there could be a "pointer" to the PBC template, for other people that would look for this? I can imagine latest after your post goes live, there might be some interest in it 😉

**Reply by nbrown02** [reply]
I think left open - some good questions you've asked that I hope others do too!

**Reply by nbrown02** [reply]
@huserben if you check the [other repo](https://github.com/nbrown02/Process-Behaviour-Chart) I just pushed out a new version which now has the ability to filter just your baseline (for UNPL/LNPL) as well as you actual data to compare, as well as the XmR chart :)

**Reply by huserben** [reply]
Well, this is some cool stuff and a great way to start the year. Thanks @nbrown02 !

**Comment by nbrown02** [comment]
Closing discussion - process behaviour charts are available in this repo - https://github.com/nbrown02/Process-Behaviour-Chart
